ipcore/vhdl_node.py

```diff
@@ -0,0 +1,373 @@
+            vhdl_file.write(self.generate())
+        
+        work_library = next((library for library in self.libraries if library.name == "work"), None)
+        cur_dir = pathlib.Path(os.path.realpath(__file__)).parent.name
+
+        if work_library is not None:
+            for package in work_library.packages:
+                if os.path.isfile(cur_dir + "/" + package + ".vhd"):
+                    copyfile(cur_dir + "/" + package + ".vhd", output_dir + package + ".vhd")
+
+class EntityFile(VHDLFile):
+    """Represents a VHDL entity file."""
+
+    def __init__(self, name, entity, architecture=None, libraries=None):
+        super().__init__(name, libraries)
+        self.entity = entity
+        self.architecture = architecture
+    def generate(self):
+        """Generate a VHDL entity file."""
+        vhdl = ""
+
+        for library in self.libraries:
+            vhdl += library.generate()
+
+        vhdl += self.entity.generate()
+        vhdl += self.architecture.generate()
+        return vhdl
+
+class Signal(BaseVHDLNode):
+    def __init__(self, name, length=1, default_value=None, data_type=None, underlying_data_type=None):
+        super().__init__(name)
+        self.length = length
+        self.default_value = default_value
+
+        if data_type is None:
+            data_type = self._get_data_type()
+        self.data_type = data_type
+        self.underlying_data_type = underlying_data_type
+    def generate(self):
+        """Generate a VHDL signal."""
+        if self.length == 1:
+            type_declaration = self.data_type
+        else:
+            type_declaration = f"{self.data_type}({(self.length - 1)} downto 0)"
+        if self.default_value is not None:
+            default_str = f" := {self._get_default_str()}"
+        else:
+            default_str = ""
+        return f"signal {self.name.ljust(32)} : {type_declaration}{default_str};\n"
+    def generate_assignment(self, new_signal_value):
+        """Generate VHDL assignment to this signal, including necessary resizing.
+
+        Resizes using underlying data type if possible.
+        VHDL data types must match except for std_logic and std_logic_vector.
+        For assignmnents between std_logic and std_logic_vector,
+        the first bit is where data is assumed to be stored.
+
+        Parameters
+        ----------
+        new_signal_value : Signal
+            signal being assigned to this. VHDL data types must match except for std_logic and std_logic_vector
+
+        Returns
+        -------
+        str
+            VHDL code to assign new_signal_value to this signal
+        """
+        if self.length == 1:
+            if new_signal_value.length == 1:
+                right_hand_side = new_signal_value.name
+            else:
+                right_hand_side = f"{new_signal_value.name}(0)"
+        elif new_signal_value.length == 1:
+            right_hand_side = f"(0 => {new_signal_value.name}, others => '0')"
```

Comment:
Just a point of clarification here: if the new value has a length of 1, it has to be unsigned in nature, so we put zeros at the front. 

I had to do a double take on these lines; I initially read the `others => '0'` and thought "but what about sign extension?" A comment here might help.